src/device/pf_cmrpc.c

```diff
@@ -2596,117 +2775,131 @@ static int pf_cmrpc_dce_packet(
+
+               if ((uint16_t)(p_sess->out_buf_len - p_sess->out_buf_sent_pos) < (PF_MAX_UDP_PAYLOAD_SIZE - start_pos))
+               {
+                  LOG_DEBUG(PF_RPC_LOG, "CMRPC(%d): Sending last fragment.\n", __LINE__);
 
-                     p_sess->out_fragment_nbr = 0;
-                  }
-                  else
-                  {
-                     LOG_DEBUG(PF_RPC_LOG, "CMRPC(%d): Sending intermediate fragment.\n", __LINE__);
-                     /* Send next fragment */
-                     /* Insert the response header to get pos of rpc response body. */
-                     pf_put_dce_rpc_header(&rpc_res, *p_res_len, p_res, &res_pos, &length_of_body_pos);
-                     start_pos = res_pos;                /* Save for later */
-
-                     /* We need to send a fragmented answer - Send an intermediate fragment now */
-                     memcpy(&p_res[res_pos], &p_sess->out_buffer[p_sess->out_buf_sent_len], *p_res_len - res_pos);
-                     p_sess->out_buf_sent_len += *p_res_len - res_pos;
-                     res_pos += *p_res_len - res_pos;
-
-                     p_sess->out_fragment_nbr++; /* Number of the next fragment */
-                  }
+                  /* We need to send one last fragment */
+                  p_sess->out_buf_send_len = p_sess->out_buf_len - p_sess->out_buf_sent_pos; /* Send all the rest */
 
-                  /* Insert the real value of length_of_body in the rpc header */
-                  pf_put_uint16(rpc_res.is_big_endian, (uint16_t)(res_pos - start_pos), *p_res_len, p_res, &length_of_body_pos);
+                  /* Send last fragment */
+                  rpc_res.flags.last_fragment = true;
+
+                  /* Update the response header to get pos of rpc response length. */
+                  pf_put_dce_rpc_header(&rpc_res, PF_MAX_UDP_PAYLOAD_SIZE, p_sess->out_buffer, &rpc_hdr_start_pos, &length_of_body_pos);
                }
                else
                {
-                  LOG_DEBUG(PF_RPC_LOG, "CMRPC(%d): No more fragments to send. Total %u, already sent %u.\n", __LINE__,
-                     p_sess->out_buf_len, p_sess->out_buf_sent_len);
-                  /* The last fragment has been acknowledged */
-                  p_sess->out_fragment_nbr = 0;
-                  res_pos = 0;  /* Nothing more to do */
+                  LOG_DEBUG(PF_RPC_LOG, "CMRPC(%d): Sending intermediate fragment.\n", __LINE__);
+
+                  /* Send next fragment */
+                  p_sess->out_buf_send_len = PF_MAX_UDP_PAYLOAD_SIZE - start_pos; /* Send as much as we can */
                }
+
+               /* Send a fragment now */
+               /* This is essentially a memmove() in order to make the next fragment data follow the RPC header at the start of out_buffer */
+               pf_put_mem(&p_sess->out_buffer[p_sess->out_buf_sent_pos], p_sess->out_buf_send_len, PF_MAX_UDP_PAYLOAD_SIZE, p_sess->out_buffer, &res_pos);
```

Comment:
This is where a memmove might be needed.

Reply:
Used the new function `pf_put_mem_overlapping() `